service/src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java

```diff
@@ -75,14 +82,78 @@
+            new LinkedBlockingQueue<Map<String, String>>();
 
     private static CountDownLatch newRequestHandlerShutdownLatch;
     private static CountDownLatch requestUpdateHandlerShutdownLatch;
     private static CountDownLatch sampleUpdateHandlerShutdownLatch;
+    private static CountDownLatch correctCmoPatientIdShutdownLatch;
     private static Gateway messagingGateway;
 
     private static final Log LOG = LogFactory.getLog(MessageHandlingServiceImpl.class);
 
+    private class CorrectCmoPatientIdReqReplyHandler implements Runnable {
+        final Phaser phaser;
+        boolean interrupted = false;
+
+        CorrectCmoPatientIdReqReplyHandler(Phaser phaser) {
+            this.phaser = phaser;
+        }
+
+        @Override
+        public void run() {
+            phaser.arrive();
+            while (true) {
+                try {
+                    Map<String, String> idCorrectionMap =
+                            correctCmoPatientIdQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if (idCorrectionMap != null) {
+                        String oldCmoPatientId = idCorrectionMap.get("oldId");
+                        String newCmoPatientId = idCorrectionMap.get("newId");
+
+                        // get samples by old cmo patient id before updating the
+                        // cmo patient id for the given patient alias/patient node
+                        List<MetadbSample> samples =
+                                sampleService.getMetadbSampleListByCmoPatientId(oldCmoPatientId);
+                        MetadbPatient updatedPatient = patientService.updateCmoPatientId(
+                                oldCmoPatientId, newCmoPatientId);
+
+                        for (MetadbSample sample: samples) {
+                            // TODO: add support for clinical sample updates
+                            // publish research samples only to label generator
+                            if (sample.getSampleCategory().equals("research")) {
+                                SampleMetadata latestMetadata = sample.getLatestSampleMetadata();
+                                latestMetadata.setCmoPatientId(newCmoPatientId);
+                                LOG.info("Replacing patient ID prefix for sample: "
+                                        + latestMetadata.getPrimaryId() + " to new CMO patient ID prefix");
+                                String newCmoSampleLabel = latestMetadata.getCmoSampleName()
+                                        .replaceAll(oldCmoPatientId, newCmoPatientId);
+                                latestMetadata.setCmoSampleName(newCmoSampleLabel);
+                                // add to sample update queue
+                                LOG.info("Persisting update for sample to database");
+                                sample.updateSampleMetadata(latestMetadata);
+                                sampleService.saveSampleMetadata(sample);
```

Comment:
Maybe we can move save and update out of the if block because clinical samples would have to be updated and saved the same way

Reply:
moved the `.save(....)` call out of the if-block